File: 프로그래머스/lv2/67257. ［카카오 인턴］ 수식 최대화/［카카오 인턴］ 수식 최대화.py
from itertools import permutations


def evaluate(expression, case, n):
    expression = expression.split(case[n])

    v = 0
    if n == 2:
        v = int(expression[0])
    else:
        v = int(evaluate(expression[0], case, n+1))

    for i in range(1, len(expression)):
        if n == 2:
            if case[n] == "*":
                v *= int(expression[i])
            elif case[n] == "+":
                v += int(expression[i])
            else:
                v -= int(expression[i])
        else:
            if case[n] == "*":
                v *= int(evaluate(expression[i], case, n+1))
            elif case[n] == "+":
                v += int(evaluate(expression[i], case, n+1))
            else:
                v -= int(evaluate(expression[i], case, n+1))
    return v


def solution(expression):
    cases = list(permutations(['*', "+", "-"], 3))

    maxAnswer = 0
    for case in cases:
        v = evaluate(expression, case, 0)
        print(abs(v))
        maxAnswer = max(maxAnswer, abs(v))

    return maxAnswer

File: 프로그래머스/2/67257. ［카카오 인턴］ 수식 최대화/［카카오 인턴］ 수식 최대화.java
public class Solution {

    static String[][] cases = {
            {"\\*", "\\+", "\\-"},
            {"\\*", "\\-", "\\+"},
            {"\\+", "\\*", "\\-"},
            {"\\+", "\\-", "\\*"},
            {"\\-", "\\*", "\\+"},
            {"\\-", "\\+", "\\*"}
    };

    public static long solution(String expression) {
        long maxAnswer = 0;
        for (String[] order : cases) {
            long value = evaluate(expression, order, 0);
            maxAnswer = Math.max(maxAnswer, Math.abs(value));
        }
        return maxAnswer;
    }

    public static long evaluate(String expression, String[] order, int n) {
        String[] tokens = expression.split(order[n]);
        long result = getValue(tokens[0], order, n);

        for (int i = 1; i < tokens.length; i++) {
            long temp = getValue(tokens[i], order, n);

            if (order[n].equals("\\*")) {
                result *= temp;
            } else if (order[n].equals("\\+")) {
                result += temp;
            } else {
                result -= temp;
            }
        }
        return result;
    }

    private static long getValue(String token, String[] order, int n) {
        if (n == 2) {
            return Long.parseLong(token);
        }
        return evaluate(token, order, n + 1);
    }
}
